test/rulesets/Base/distributed.jl

```diff
@@ -0,0 +1,12 @@
+@testset "pmap" begin
+    test_rrule(pmap, inv, WorkerPool(), rand(10), check_inferred=false) # test empty worker pool
+    test_rrule(pmap, inv, default_worker_pool(), rand(10), check_inferred=false)
+    test_rrule(pmap, inv, default_worker_pool(), rand(4, 4), check_inferred=false) # test matrix input
+    test_rrule(pmap, inv, default_worker_pool(), rand(10), fkwargs=(batch_size=2,), check_inferred=false) # test batch_size > 1
+    test_rrule(pmap, Multiplier(2.0), default_worker_pool(), rand(3), check_inferred=false) # test adjoint of f
```

Comment:
There's a lot of pretty complicated code. Do these tests cover a reasonable list of edge cases? Like:
* What can be used instead of default_worker_pool()? Would they notice if this was wrong?
* Does it matter whether batch sizes divide the number of workers evenly? If `addprocs()` is called somewhere?
* Will it obey pmap's keywords? `distributed=false`, `retry_delays`?
* What functions `f` are allowed, can I `pmap` a Flux model? It can't guarantee an order, which places some restrictions. Can it give an error if these are violated?
* Or how about a function which isn't type-stable, or worse, sometimes gives `NoTangent()`? Will `similar(res_batches[1], output_sz)` get confused?
* Does this only work on arrays of numbers? Then it should not accept `X::Any`.

Reply:
- Any worker pool can be used; default_worker_pool is the default for `pmap` when none is provided.
- That doesn't matter; the rule just remembers the mapping that `pmap` chose to use in the forward pass and does the same in the backward pass. Here is a diagram:
![IMG_0325](https://user-images.githubusercontent.com/43899491/159338554-371bb1c7-3120-4438-a776-2ffc6b4a755d.jpg)
- I added a test for `distributed=false`. It won't error with `retry_delays`, but it won't do retries in the backward pass, which could be a problem, I'm not sure about those sorts of situations. In general, this rule makes assumptions about the robustness of worker process, e.g. that they don't die between the forward and backward pass
- Any function `f` should work
- It could; I don't think there's actually any point in aiming for type stability in there, so I should change that. Going to hold off on it for now because I think there may be bigger issues with projection in that last part
- The aim is to work for arbitrary collections. In particular, note that `pmap` handles multiple collections by calling the single-collection version with all the collections zipped. By writing this rule for an arbitrary collection, I especially wanted to handle zipped ararys.

Reply:
Re 5, I don't mean type-stability for performance, I mean that you may create a container which doesn't hold what it needs to. Which is why you need to test with an `f` whose gradient is sometimes zero. 

Re 4, any `f`: Must it be defined `@everywhere`? Can it close over arrays defined only locally? 

Re 1, if "Any worker pool can be used", then ideally you would test that it does in fact use the pool specified, both forward and backward.

Reply:
- You're right, I think that could definitely cause a bug.
- It can be any `f` that is valid for use in `pmap`.
- It does more than just using the same pool, it uses the exact same matching of workers to tasks.
- Sure, I could explicitly check that, although this isn't something that's really in danger of happening since it explicitly uses precisely the same IDs as the forward pass.